collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright 2018 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android;
+
+import android.content.res.AssetManager;
+import android.os.Environment;
+import android.support.test.InstrumentationRegistry;
+import android.support.test.runner.AndroidJUnit4;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.listeners.FormLoaderListener;
+import org.odk.collect.android.logic.FormController;
+import org.odk.collect.android.tasks.FormLoaderTask;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import timber.log.Timber;
+
+import static junit.framework.Assert.assertEquals;
+
+@RunWith(AndroidJUnit4.class)
+public class FormNavigationTestCase {
+
+    private static final String FORMS_DIRECTORY = "/odk/forms/";
+
+    @Test
+    public void test1() throws IOException {
+        String formName = "form1.xml";
+        prepareFile(formName);
+
+        FormLoaderTask formLoaderTask = new FormLoaderTask(formPath(formName), null, null);
+        formLoaderTask.setFormLoaderListener(new FormLoaderListener() {
+            @Override
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    FormController formController = task.getFormController();
+                    assertEquals("-1, ", formController.getFormIndex().toString());
+                    formController.stepToNextScreenEvent();
+                    assertEquals("0, ", formController.getFormIndex().toString());
+                    formController.stepToNextScreenEvent();
+                    assertEquals("-1, ", formController.getFormIndex().toString());
+                    formController.stepToPreviousScreenEvent();
+                    assertEquals("0, ", formController.getFormIndex().toString());
+                } catch (Exception e) {
+                    Timber.i(e);
+                }
+            }
+
+            @Override
+            public void loadingError(String errorMsg) {
+            }
+
+            @Override
+            public void onProgressStep(String stepMessage) {
+
+            }
+        });
+        formLoaderTask.execute(formPath(formName));
+    }
+
+    @Test
```

Comment:
There are 390 lines of code spent in this file. How about factoring out the identical parts of these tests, or combining them in some other way.

“My point today is that, if we wish to count lines of code, we should not regard them as "lines produced" but as "lines spent": the current conventional wisdom is so foolish as to book that count on the wrong side of the ledger.”
Dijkstra (1988) "On the cruelty of really teaching computing science (EWD1036).